Modules/Core/Common/src/itkObjectFactoryBase.cxx

```diff
@@ -237,14 +238,19 @@ ObjectFactoryBase::Initialize()
 {
   itkInitGlobalsMacro(PimplGlobals);
 
-  // Atomically set m_Initialized to true. If it was false before, enter the if.
-  if (!m_PimplGlobals->m_Initialized.exchange(true))
+  if (!m_PimplGlobals->m_Initialized)
   {
-    ObjectFactoryBase::InitializeFactoryList();
-    ObjectFactoryBase::RegisterInternal();
+    // only one thread initializing at a time, wait until initialization is completed.
+    std::lock_guard lockGuard(m_PimplGlobals->m_Lock);
+    if (!m_PimplGlobals->m_Initialized)
```

Comment:
Looks like you're doing _double-checked locking_, which is considered an _anti-pattern_ nowadays! If one thread sets `m_PimplGlobals->m_Initialized = true` while at the same time, another thread tries to evaluate `m_PimplGlobals->m_Initialized` outside the locked region, we have undefined behavior.

Reply:
What is the updated pattern? To ensure it's only initialized once and block other threads until it it initialed?

Reply:
std::call_once?

Reply:
My bigger concern now is if using `std::mutex` during static initialization is OK aka ( defined behavior ).

> evaluate m_PimplGlobals->m_Initialized outside the locked region, we have undefined behavior.

Not sure how undefined the undefined read behavior would be, maybe just not 100% synced between threads. It's will either be true or false, either will be OK.

Reply:
> std::call_once?

Thanks, @dzenanz. I just gave it a try, using std::call_once. Please check! (Still draft!)
- #4227